server/src/main/scala/org/http4s/server/middleware/Logger.scala

```diff
@@ -19,32 +19,38 @@ object Logger {
       )
     )
 
-  def getCharsetFromContentType(m: Message): (Boolean, Option[Charset]) = {
-    val init = m.charset
-    val binary = m.contentType.map(_.mediaType.binary).getOrElse(false)
-    (binary, init)
-  }
 
-  /**
-    * Log Messages As They Pass
-    */
-  def messageLogPipe[A <: Message](logHeaders: Boolean, logBody: Boolean)
-                                  (logger: SLogger): Pipe[Task, A, A] = stream => {
-    stream.flatMap { req =>
-      val (binary, charset) = Logger.getCharsetFromContentType(req)
-      val headers = if (logHeaders) req.headers.toList.mkString("Headers(", ", ", ")") else ""
-      val bodyT: Task[String] = {
-        if (logBody && !binary) {
-          req.bodyAsText(charset.getOrElse(Charset.`UTF-8`))
-        } else if (logBody) {
-          req.body.map(ByteVector.fromByte).map(_.toHex)
-        } else {
-          Stream.empty[Task, String]
-        }
-      }.runFoldMap(identity)
-      Stream.eval(bodyT.flatMap(body =>
-        Task.delay(logger.info(s"$headers + Body: $body"))
-      )) >> Stream.emit(req)
+  def logMessage[A <: Message](message: A)
+                              (logHeaders: Boolean, logBody: Boolean)
+                              (logger: SLogger)
+                              (implicit strategy: Strategy): Task[Unit] = {
+
+    val charset = message.charset
+    val binary = message.contentType.exists(_.mediaType.binary)
+
+    val headers = if (logHeaders) message.headers.toList.mkString("Headers(", ", ", ")") else ""
```

Comment:
The only problem I saw is that this would be very verbose assuming you were only logging Headers, which is why I opted the direction I did. I was attempting to make the logging close to idiomatic scala output, but could be easily convinced to change.